### PR TITLE
Initial terraform automation

### PR DESCRIPTION
**Pull request by rancher-max**
#### Proposed Changes ####

Add terraform automation to rke2. This should allow users to create rke2 clusters in AWS easily, and in a repeatable fashion, using the terraform here. It also moves our terraform automation from the rancher/rancher Python framework directly here in the repo it belongs.

#### Types of Changes ####

New feature primarily for testing

#### Verification ####

See README included in this PR for how to run tests.

#### Linked Issues ####
N/A

#### Further Comments ####

I'm not a Go expert, so the Go code is likely really bad. Please feel free to let me know what to do better here and I can update as needed.

I've confirmed that the Jenkinsfile and build scripts provided succeed when running this in Jenkins. The output initially has a lot of terraform details for what it is doing, then gives the KUBECONFIG, followed by a detailed view of the cluster's nodes and pods (`kubectl get nodes,pods -A -o wide`), followed by individual validation tests:
```
• [SLOW TEST] [83.309 seconds]
Test: Build Cluster: Checks Node and Pod Status
/go/src/github.com/rancher/rke2/tests/terraform/createcluster_test.go:45
------------------------------
• [SLOW TEST] [40.509 seconds]
Test: Build Cluster: Verifies ClusterIP Service
/go/src/github.com/rancher/rke2/tests/terraform/createcluster_test.go:86
------------------------------
• [SLOW TEST] [21.248 seconds]
Test: Build Cluster: Verifies NodePort Service
/go/src/github.com/rancher/rke2/tests/terraform/createcluster_test.go:115
------------------------------
• [SLOW TEST] [70.654 seconds]
Test: Build Cluster: Verifies Ingress
/go/src/github.com/rancher/rke2/tests/terraform/createcluster_test.go:147
------------------------------
• [SLOW TEST] [19.537 seconds]
Test: Build Cluster: Verifies Daemonset
/go/src/github.com/rancher/rke2/tests/terraform/createcluster_test.go:185
------------------------------
• [SLOW TEST] [49.295 seconds]
Test: Build Cluster: Verifies dns access
/go/src/github.com/rancher/rke2/tests/terraform/createcluster_test.go:204
------------------------------
PASSED!

Ran 7 of 7 Specs in 1485.964 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: Test_TFClusterCreateValidation (1485.97s)
PASS
ok  	command-line-arguments	1486.003s
```

### CODE REVIEW

File: tests/terraform/cluster.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	varDir, err := filepath.Abs(basepath + tfVarsPath)
+	if err != nil {
+		return "", err
+	}
+	TerraformOptions := &terraform.Options{
+		TerraformDir: tfDir,
+		VarFiles:     []string{varDir},
+	}
+	NumServers, _ = strconv.Atoi(terraform.GetVariableAsStringFromVarFile(t, varDir, "no_of_server_nodes"))
```
Comment (briandowns) [inline tests/terraform/cluster.go:37]: Check the error here and anywhere else ignored.

File: tests/terraform/cluster.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	basepath := GetBasepath()
+	tfDir, err := filepath.Abs(basepath + "/tests/terraform/modules")
+	if err != nil {
+		return "", err
+	}
+	varDir, err := filepath.Abs(basepath + tfVarsPath)
+	if err != nil {
+		return "", err
+	}
+	TerraformOptions := &terraform.Options{
```
Comment (briandowns) [inline tests/terraform/cluster.go:33]: Does this need to be exported?

File: tests/terraform/cluster.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		terraform.Destroy(t, TerraformOptions)
+		return "cluster destroyed", err
+	}
+
+	fmt.Printf("Creating Cluster")
+	terraform.InitAndApply(t, TerraformOptions)
+	KubeConfigFile = terraform.Output(t, TerraformOptions, "kubeconfig")
+	MasterIPs = terraform.Output(t, TerraformOptions, "master_ips")
+	WorkerIPs = terraform.Output(t, TerraformOptions, "worker_ips")
+	return "cluster created", err
```
Comment (briandowns) [inline tests/terraform/cluster.go:62]: Where did this error come from? Check and return or exit. Letting this error live unnecessarily is a code smell.

File: tests/terraform/createcluster_test.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var tfVars = flag.String("tfvars", "/tests/terraform/modules/config/local.tfvars", "custom .tfvars file from base project path")
+var destroy = flag.Bool("destroy", false, "a bool")
+
+var failed = false
```
Comment (briandowns) [inline tests/terraform/createcluster_test.go:17]:
```suggestion
var failed bool
```

File: tests/terraform/createcluster_test.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+		It("Checks Node and Pod Status", func() {
+			defer func() {
+				_, err := GetNodes(KubeConfigFile, true)
+				if err != nil {
+					fmt.Println("Error retrieving nodes: ", err)
+				}
+				_, err = GetPods(KubeConfigFile, true)
+				if err != nil {
+					fmt.Println("Error retrieving pods: ", err)
```
Comment (briandowns) [inline tests/terraform/createcluster_test.go:53]: Should we exit here if pods can't be gotten?

File: tests/terraform/testutils.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+		}
+		pods = append(pods, pod)
+	}
+	if print {
+		fmt.Println(podList)
+	}
+	return pods, nil
+}
+
+func GetPods(kubeconfig string, print bool) ([]Pod, error) {
```
Comment (briandowns) [inline tests/terraform/testutils.go:269]: Using "get" in Go code isn't idiomtic and a left over from other languages. The more common and idiomatic way is to name the function Pods as that's what is returns. Same for all of the other "get" functions.

File: tests/terraform/testutils.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+		ingressIPs := strings.Split(ingressIP, " ")
+		return ingressIPs, nil
+	} else {
+		return nil, nil
+	}
+}
+
+func parseNodes(kubeConfig string, print bool, cmd string) ([]Node, error) {
+	nodes := make([]Node, 0, 10)
+	nodeList := ""
```
Comment (briandowns) [inline tests/terraform/testutils.go:205]:
```suggestion
	var nodes string
```

Comment (briandowns) [inline tests/terraform/testutils.go:205]: Since Go is statically and strongly typed, it's an antipattern to use Hungarian notation.

File: tests/terraform/createcluster_test.go (hunk 5 above)
Comment (rancher-max) [reply tests/terraform/createcluster_test.go:53]: My goal is for this to print the pods at the very end of this test. The test will fail in the block below:
```

			Eventually(func(g Gomega) {
				pods, err := GetPods(kubeConfigFile, false)
				g.Expect(err).NotTo(HaveOccurred())
				for _, pod := range pods {
					if strings.Contains(pod.Name, "helm-install") {
						g.Expect(pod.Status).Should(Equal("Completed"), pod.Name)
					} else {
						g.Expect(pod.Status).Should(Equal("Running"), pod.Name)
						g.Expect(pod.Restarts).Should(Equal("0"), pod.Name)
						numRunning := re.FindAllString(pod.Ready, 2)
						g.Expect(numRunning[0]).Should(Equal(numRunning[1]), pod.Name, "should have all containers running")
					}
				}
			}, "600s", "5s").Should(Succeed())
```
but this specific bit on lines 51-53 is more to just print for visibility so we see what is (or isn't) running at the end of this test.

File: tests/terraform/cluster.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/terraform"
+)
+
+var (
+	KubeConfigFile string
```
[the remark this answers is not in the document]
Comment (rancher-max) [reply tests/terraform/cluster.go]: Did both the variables and the function names everywhere

File: tests/terraform/createcluster_test.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+var _ = AfterEach(func() {
+	failed = failed || CurrentSpecReport().Failed()
+})
+
+var _ = BeforeEach(func() {
+	failed = failed || CurrentSpecReport().Failed()
+	if *destroy {
+		Skip("Cluster is being Deleted")
+	}
+})
```
Comment (dereknola) [inline tests/terraform/createcluster_test.go:218]: What is this section trying to achieve? Are you simply trying to check for errors and then skip to the end if one is found?

Comment (rancher-max) [reply tests/terraform/createcluster_test.go:218]: This is a good callout...
To be honest, the `AfterEach` can probably be removed entirely, and the update to the `failed` variable in the `BeforeEach` can be also, since they're just there for this pretty irrelevant block at the end of the suite:
```
	if failed {
		fmt.Println("FAILED!")
	} else {
		fmt.Println("PASSED!")
	}
```
The `destroy` block in the `BeforeEach` is there primarily for local runs of this. It purposely skips all the specs and just destroys all the resources created by terraform at the end of the suite:
```
	if *destroy {
		status, err := buildCluster(&testing.T{}, *tfVars, *destroy)
		Expect(err).NotTo(HaveOccurred())
		Expect(status).To(Equal("cluster destroyed"))
	}
```

TLDR: I'll remove the unnecessary bits 👍

**Review by dereknola — APPROVED** [review]

**Comment by rancher-max** [comment]
Latest push was to rebase off master and resolve conflicts in go.mod. This is ready for review again.

File: tests/terraform/modules/join_rke2_master.sh [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+hostname=$(hostname -f)
+mkdir -p /etc/rancher/rke2
+cat <<EOF >>/etc/rancher/rke2/config.yaml
+write-kubeconfig-mode: "0644"
+tls-san:
+  - ${create_lb}
+server: https://${initial_node_ip}:9345
+token:  "${token}"
+node-name: "${hostname}"
```
Comment (VestigeJ) [inline tests/terraform/modules/join_rke2_master.sh:27]: should we support more exhaustive config.yaml options?
ie 

cni: cilium

Comment (rancher-max) [reply tests/terraform/modules/join_rke2_master.sh:27]: This should cover that! It accepts a `.tfvars` file, and in there we have server and agent args, so passing something like:
`cni: cilium\nprofile: cis-1.6"` should work for both server and agent and add to this config.yaml. That should be present on line 33 from this file:
```
echo -e "$server_flags" >> /etc/rancher/rke2/config.yaml
```

**Review by VestigeJ — APPROVED** [review]
LargeLGTM If there are logical or syntactical errors they're above my cursory reading level to catch

File: tests/terraform/cluster.go [hunk 11]
```diff
@@ -11,53 +10,73 @@ import (
...
 
 var (
-	KubeConfigFile string
-	MasterIPs      string
-	WorkerIPs      string
-	NumServers     int
-	NumWorkers     int
-	AwsUser        string
-	AccessKey      string
+	kubeConfigFile string
```
[the remark this answers is not in the document]
Comment (rancher-max) [reply tests/terraform/cluster.go:13]: Yep these ones are used in multiple tests from `createcluster_test.go` but created in this `cluster.go`

**Review by briandowns — APPROVED** [review]
